app/src/main/java/de/westnordost/streetcomplete/overlays/surface/PathSurfaceOverlay.kt

```diff
@@ -0,0 +1,206 @@
+import de.westnordost.streetcomplete.overlays.Style
+import de.westnordost.streetcomplete.quests.surface.AddPathSurface
+import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.BICYCLIST
+import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.OUTDOORS
+import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.PEDESTRIAN
+import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.WHEELCHAIR
+import de.westnordost.streetcomplete.osm.sidewalk_surface.createSidewalkSurface
+import de.westnordost.streetcomplete.osm.surface.SurfaceAndNote
+
+class PathSurfaceOverlay : Overlay {
+
+    override val title = R.string.overlay_path_surface
+    override val icon = R.drawable.ic_quest_way_surface
+    override val changesetComment = "Specify path surfaces"
+    override val wikiLink: String = "Key:surface"
+    override val achievements = listOf(PEDESTRIAN, WHEELCHAIR, BICYCLIST, OUTDOORS)
+    override val hidesQuestTypes = setOf(AddPathSurface::class.simpleName!!)
+
+    private val handledSurfaces = Surface.values().map { it.osmValue }.toSet() + INVALID_SURFACES
+
+    override fun getStyledElements(mapData: MapDataWithGeometry): Sequence<Pair<Element, Style>> {
+        return mapData
+           .filter( """
+               ways, relations with
+               (
+                   highway ~ ${(ALL_PATHS).joinToString("|")}
+                   and (!surface or surface ~ ${handledSurfaces.joinToString("|") })
+                   and (!cycleway:surface or cycleway:surface ~ ${handledSurfaces.joinToString("|") })
+                   and (!footway:surface or footway:surface ~ ${handledSurfaces.joinToString("|") })
+                   and (segregated = yes or (!cycleway:surface and !footway:surface))
+                   and (!surface:note or (surface or cycleway:surface or footway:surface or segregated = yes))
+                   and (!cycleway:surface:note or cycleway:surface)
+                   and (!footway:surface:note or footway:surface)
+                   and !sidewalk and !sidewalk:left and !sidewalk:right and !sidewalk:both
+                   and !sidewalk:both:surface and !sidewalk:right:surface and !sidewalk:left:surface and !sidewalk:surface
+                   and !sidewalk:both:surface:note and !sidewalk:right:surface:note and !sidewalk:left:surface:note
+               )
+               or
+               (
+                   highway ~ ${(ALL_ROADS).joinToString("|")}
+                   and (sidewalk ~ left|right|both or sidewalk:both = yes or sidewalk:left = yes or sidewalk:right = yes)
+                   and (!sidewalk:both:surface or sidewalk:both:surface ~ ${handledSurfaces.joinToString("|") })
+                   and (!sidewalk:right:surface or sidewalk:right:surface ~ ${handledSurfaces.joinToString("|") })
+                   and (!sidewalk:left:surface or sidewalk:left:surface ~ ${handledSurfaces.joinToString("|") })
```

Comment:
Maybe it would be easier and less error-prone (I just fixed a bug in the last commit) if the filter of styled elements would be done completely separate, like for the sidewalk overlay (two `mapData.filter. ....` appended to each other):

- instead of putting the sidewalk check into the expression, you could just use the parser functions to check, e.g. `createSidewalkSides`, `createSidewalkSurface` etc..
- no need to filter twice, i.e. first filter elements and then assume in `getStyleForSidewalkAsProperty` that it is a road with sidewalk. Instead, you can parse the sidewalk surface once and then return maybe null for the style if it has no sidewalk after all
- no additional switch at `getStyle`

Reply:
Sounds like a good idea.

One thing that worries me in general is that I have no real idea about performance implications - I have no idea whether either version will be faster and is it change that matters. I just keep assuming that it will be negligible. 

I case of adding some quests that were doing more complex geometry checking I looked a bit in timing logs, but I mostly confirmed that these were not atrociously slow.

Reply:
It will likely be negligibly faster because the parsing is only done once. But the difference will be somewhat negligible.

Reply:
changed

Reply:
Hm, this is not entirely what I meant but anyway a first step. I meant e.g.

filter all things that can have a sidewalk -> mapNotNull with getStyleForRoad. That function just returns null if `createSidewalkSides` or whatever was the name of that parser function returns null.

etc.. I may do this myself if you couldn't follow this explanation.

Reply:
@westnordost 

I am confused.

`.mapNotNull { if(it.id > 10000) null else it to getStyleForSidewalkAsProperty(it) }` with `getStyleForSidewalkAsProperty` returning `PolylineStyle` works fine and Kotlin detects that it will return only nonnulls.

With `getStyleForSidewalkAsProperty` returning `PolylineStyle` it suddenly complains that type is mismatching because it can return nulls.

Tried again, run into the same. Not sure whether I am making some dumb mistake or is it Kotlin being silly.